daml/ContingentClaims/Util.daml

```diff
@@ -80,3 +83,22 @@ pruneZeros' other = embed other
 -- | This avoids requiring the equality type constraint on `a`
 isZero Zero = True
 isZero _ = False
+
+-- | Returns the claim's implied acquisition time (IAT) if it exists.
+-- It is defined as the time `t` such that, for the purpose of lifecycling and valuation, it is irrelevant whether the contract has been acquired at `t` or at any time `s` before `t`.
+-- This check is required as currently the claim's acquisition time is not an input to `lifecycle`, which forces us to work with claims that have an IAT.
+-- For instance: 
+-- - `Scale obs c` does not have an IAT (as `obs` is evaluated at the claim's acquisition time)
+-- - the IAT for `When (TimeGte t) c` is `t`
+-- - `When (obs <= 100) c` does not have an IAT, as the acquisition time defines when we start to monitor the observable `obs`
+impliedAcquisitionTime : (Ord t) => Claim t x a -> Optional t
+impliedAcquisitionTime (Anytime (TimeGte t) _) = Some t
+impliedAcquisitionTime (When (TimeGte t) _) = Some t
+impliedAcquisitionTime (And c1 c2 cs) = foldr folder (impliedAcquisitionTime c1) (c2 :: cs)
+  where folder = flip (liftA2 min) . impliedAcquisitionTime
+impliedAcquisitionTime _ = None
```

Comment:
I suggested during your presentation using a `HasTime` constraint to make this function total. Have you considered that?

Perhaps it may not be enough on it's own (what happens if I want to evaluate a function in the future/past), but if you have `t`, as in the `lifecycle` function, maybe take the limit of that and `getTime`? Would that work?

I guess what I'm saying is that I don't like the fact that we are admitting undefined cases. Not in this particular funciton, but up above you fail if you don't have an acquisition time. Is it really a degenerate case, or can we pick some reasonable value for the acquisition time, as I suggested above, and use that?
I'm not sure whether it's preferable to fail fast or to have this 'caveat' for degenerate cases. What do you think?

Reply:
Thanks for the feedback. 

For now I am keeping a strict separation between the time used to lifecycle claims and ledger time, hence I am a bit reluctant to leverage `HasTime` (which, if I am not mistaken, would allow us to fetch ledger time within an `Update` monad.

Regarding the usage of an undefined case, my take is the following.

An acquisition time is defined for each claim: if we enter into a forward contract today, the today date would be a reasonable acquisition time for the claim.

We would have
- `c = When ( t >= T ) ( payoff ) `
- `Acquisition Time = today`

However, our lifecycle API does not take acquisition time as an input, so we can't use this information to lifecycle the contract

The next best thing we can do is to recognize that for some claims, such as the one above, we do not need the actual acquisition time in order to properly lifecycle the contract. I called this feature "having an intrinsic acquisition time",  meaning that we can assume that `Acquisition Time = T` when traversing the tree.

Other claims, e.g. `c = Scale S(t) payoff`, cannot be handled properly by our `lifecycle` function as we would need the exact acquisition time in order to calculate the payment amount. In this case, I would rather avoid making any assumptions and throw an error. 

As I mentioned on Friday, I think that one day we can take the chance to change the signature of `lifecycle` to include the claim's acquisition time. Then we can remove the intrinsic acquisition time business altogether.

TL,DR : I don't think we are admitting undefined cases (the acquisition time is always defined) but at the moment the `lifecycle` function does not have enough information to handle all cases